src/main/java/org/broadinstitute/hellbender/utils/dataflow/SmallBamWriter.java

```diff
@@ -75,4 +98,71 @@ public void processElement(ProcessContext c) throws Exception {
+            return;
+        }
+
+        String headerString = Base64.getEncoder().encodeToString(SerializableUtils.serializeToByteArray(header));
+
+        @SuppressWarnings("unchecked")
+        Class<? extends FileOutputFormat<NullWritable, SAMRecordWritable>> outputFormatClass =
+                (Class<? extends FileOutputFormat<NullWritable, SAMRecordWritable>>) (Class<?>) TemplatedKeyIgnoringBAMOutputFormat.class;
+        @SuppressWarnings("unchecked")
+        HadoopIO.Write.Bound<NullWritable, SAMRecordWritable> write = HadoopIO.Write.to(destPath,
+                outputFormatClass, NullWritable.class, SAMRecordWritable.class)
+                .withConfigurationProperty(TemplatedKeyIgnoringBAMOutputFormat.SAM_HEADER_PROPERTY_NAME, headerString);
+
+        reads.apply(ParDo.of(new DoFn<GATKRead, KV<NullWritable, SAMRecordWritable>>() {
+            private static final long serialVersionUID = 1L;
+            @Override
+            public void processElement(ProcessContext c) throws Exception {
+                SAMRecord samRecord = c.element().convertToSAMRecord(header);
+                SAMRecordWritable samRecordWritable = new SAMRecordWritable();
+                samRecordWritable.set(samRecord);
+                c.output(KV.of(NullWritable.get(), samRecordWritable));
+            }
+        })).setCoder(KvCoder.of(WritableCoder.of(NullWritable.class), WritableCoder.of(SAMRecordWritable.class)))
+                .apply(write.withoutSharding()); // write as a single (unsharded) file
+    }
+
+    public static class TemplatedKeyIgnoringBAMOutputFormat<K> extends KeyIgnoringBAMOutputFormat<K>
+        implements Configurable, ShardNameTemplateAware {
+
+        public static final String SAM_HEADER_PROPERTY_NAME = "sam.header";
+
+        private Configuration conf;
+
+        @Override
+        public void setConf(Configuration conf) {
+            this.conf = conf;
+            if (conf != null) {
+                String headerString = conf.get(SAM_HEADER_PROPERTY_NAME);
+                if (headerString == null) {
+                    throw new IllegalStateException("SAM file header has not been set");
+                }
+                byte[] headerBytes = Base64.getDecoder().decode(headerString);
+                setSAMHeader((SAMFileHeader) SerializableUtils.deserializeFromByteArray(headerBytes, "SAMFileHeader"));
+            }
+        }
+
+        @Override
+        public Configuration getConf() {
+            return getConf();
+        }
+
+        @Override
+        public void checkOutputSpecs(JobContext job) {
+            // don't fail if the output already exists
+        }
+
+        @Override
+        public Path getDefaultWorkFile(TaskAttemptContext context,
+                String extension) throws IOException {
+            // note that the passed-in extension is ignored since it comes from the template
+            return ShardNameTemplateHelper.getDefaultWorkFile(this, context);
```

Comment:
Can you add a little more to the comment here. What's exactly happening in hardNameTemplateHelper.getDefaultWorkFile? It's in a different repo, so it's worth adding more context.